File: ocaml/day2/main.ml
open Core
open Types

let input = In_channel.read_lines "../inputs/day2.txt" |> List.map ~f:to_round
let total_score rounds = List.map ~f:score rounds |> List.fold ~init:0 ~f:( + )

let total_score_with_strategy rounds =
  List.map ~f:score_with_strategy rounds |> List.fold ~init:0 ~f:( + )

let () =
  printf "Task 1: %d\n" @@ total_score input;
  printf "Task 2: %d\n" @@ total_score_with_strategy input

File: rust/day2/src/main.rs
mod types;

use std::fs::File;
use std::io::{BufRead, BufReader, Result as IOResult};
use types::*;

const INPUT_PATH: &str = "../inputs/day2.txt";

fn read_rounds() -> IOResult<Vec<Round>> {
    Ok(BufReader::new(File::open(INPUT_PATH)?)
        .lines()
        .map(|line| {
            line.expect("Lines are formatted properly")
                .split_whitespace()
                .map(Shape::from)
                .collect::<Vec<_>>()
        })
        .map(|round_vec| (round_vec[0], round_vec[1]))
        .collect())
}

fn total_score_faceoff(rounds: &Vec<Round>) -> Score {
    rounds
        .iter()
        .fold(0, |score, &round| score + get_score(round))
}

fn total_score_strategy(rounds: &Vec<Round>) -> Score {
    rounds
        .iter()
        .map(|&(them, you)| (them, Outcome::from(you)))
        .fold(0, |score, inst| score + play_instruction(inst))
}

fn main() -> IOResult<()> {
    let rounds = read_rounds()?;
    println!("Task 1: {}", total_score_faceoff(&rounds));
    println!("Task 2: {}", total_score_strategy(&rounds));
    Ok(())
}
